source/common/router/header_parser.cc

```diff
@@ -1,28 +1,206 @@
 #include "common/router/header_parser.h"
 
+#include <ctype.h>
+
+#include <string>
+
+#include "common/common/assert.h"
 #include "common/protobuf/utility.h"
 
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+
 namespace Envoy {
 namespace Router {
 
 namespace {
 
+enum class ParserState {
+  Literal,                   // processing literal data
+  VariableName,              // consuming a %VAR% name
+  ExpectArray,               // expect starting [ in %VAR([...])%
+  ExpectString,              // expect starting " in array of strings
+  String,                    // consuming an array element string
+  ExpectArrayDelimiterOrEnd, // expect array delimiter (,) or end of array (])
+  ExpectArgsEnd,             // expect closing ) in %VAR(...)%
+  ExpectVariableEnd          // expect closing % in %VAR(...)%
+};
+
+std::string unescape(absl::string_view sv) { return absl::StrJoin(absl::StrSplit(sv, "%%"), "%"); }
```

Comment:
It's worth a comment at least why you didn't use absl::StrReplaceAll().

Reply:
I'll switch to that function. It's not on their docs website (that I saw), so I didn't realize it existed.